src/cloudflare/internal/ai-api.ts

```diff
@@ -0,0 +1,124 @@
+// Copyright (c) 2024 Cloudflare, Inc.
+// Licensed under the Apache 2.0 license found in the LICENSE file or at:
+//     https://opensource.org/licenses/Apache-2.0
+
+interface Fetcher {
+  fetch: typeof fetch
+}
+
+export type SessionOptions = {  // Deprecated, do not use this
+  extraHeaders?: object;
+};
+
+export type AiOptions = {
+  debug?: boolean;
+  prefix?: string;
+  extraHeaders?: object;
+  sessionOptions?: SessionOptions;
+};
+
+export class InferenceUpstreamError extends Error {
+  public httpCode: number;
+
+  public constructor(message: string, httpCode: number) {
+    super(message);
+    this.name = "InferenceUpstreamError";
+    this.httpCode = httpCode;
+  }
+}
+
+export class Ai {
+  private readonly fetcher: Fetcher
+
+  private options: AiOptions = {};
+  private logs: Array<string> = [];
+  public lastRequestId: string | null = null;
+
+  public constructor(fetcher: Fetcher) {
+    this.fetcher = fetcher
+  }
+
+  public async fetch(input: RequestInfo | URL, init?: RequestInit): Promise<Response> {
+    return this.fetcher.fetch(input, init)
+  }
+
+  public async run(
+    model: string,
+    inputs: Record<string, object>,
+    options: AiOptions = {}
+  ): Promise<Response | ReadableStream<Uint8Array> | object | null> {
+    this.options = options;
+    this.lastRequestId = "";
+
+    const body = JSON.stringify({
+      inputs,
+      options: {
+        debug: this.options?.debug,
+      },
+    });
+
+    const fetchOptions = {
+      method: "POST",
+      body: body,
+      headers: {
+        ...(this.options?.sessionOptions?.extraHeaders || {}),
+        ...(this.options?.extraHeaders || {}),
+        "content-encoding": "application/json",
+        // 'content-encoding': 'gzip',
+        "cf-consn-sdk-version": "2.0.0",
+        "cf-consn-model-id": `${this.options.prefix ? `${this.options.prefix}:` : ""}${model}`,
+      },
+    };
+
+    const res = await this.fetcher.fetch("http://workers-binding.ai/run?version=2", fetchOptions);
+
+    this.lastRequestId = res.headers.get("cf-ai-req-id");
+
+    if (inputs['stream']) {
+      if (!res.ok) {
+        throw new InferenceUpstreamError(await res.text(), res.status);
+      }
+
+      return res.body;
+    } else {
+      // load logs
+      if (this.options.debug) {
+        let parsedLogs: string[] = [];
+        try {
+          const logHeader = res.headers.get("cf-ai-logs")
+          if (logHeader) {
+            parsedLogs = (JSON.parse(atob(logHeader)) as string[]);
+          }
+        } catch (e) {
+          /* empty */
+        }
+
+        this.logs = parsedLogs;
+      }
+
+      if (!res.ok || !res.body) {
+        throw new InferenceUpstreamError(await res.text(), res.status);
+      }
+
+      // Non streaming responses are always in gzip
+      // eslint-disable-next-line
+      const decompressed = new Response(res.body.pipeThrough(new DecompressionStream("gzip")));
```

Comment:
Why create a separate `Response` here? And is passing it through `DecompressionStream` necessary? If the transfer encoding is identified correctly does the `res.body` already automatically decompress the gzip?

Reply:
didn't know fetch would automatically decompress the gzip based on that header, but tried it now locally with the header on the server and it worked fine 👍 
fixed